components/validator/component.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/runtime/event"
+	"github.com/iotaledger/hive.go/runtime/timed"
+	"github.com/iotaledger/inx-app/pkg/nodebridge"
+	"github.com/iotaledger/inx-validator/pkg/daemon"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+func init() {
+	Component = &app.Component{
+		Name:     "Validator",
+		DepsFunc: func(cDeps dependencies) { deps = cDeps },
+		Params:   params,
+		Run:      run,
+		Provide:  provide,
+	}
+}
+
+var (
+	Component *app.Component
+	deps      dependencies
+
+	isValidator      atomic.Bool
+	executor         *timed.TaskExecutor[TaskType]
+	validatorAccount Account
+)
+
+type dependencies struct {
+	dig.In
+
+	NodeBridge      *nodebridge.NodeBridge
+	AccountAddress  *iotago.AccountAddress
+	PrivateKey      ed25519.PrivateKey
+	ShutdownHandler *shutdown.ShutdownHandler
+}
+
+func provide(c *dig.Container) error {
+	type depsIn struct {
+		dig.In
+		NodeBridge *nodebridge.NodeBridge
+	}
+
+	if err := c.Provide(func(deps depsIn) (*iotago.AccountAddress, error) {
+		if ParamsValidator.AccountAddress == "" {
+			return nil, ierrors.Errorf("empty bech32 in config")
+		}
+
+		hrp, addr, err := iotago.ParseBech32(ParamsValidator.AccountAddress)
+		if err != nil {
+			return nil, ierrors.Wrapf(err, "invalid bech32 address: %s", ParamsValidator.AccountAddress)
+		}
+
+		if deps.NodeBridge.APIProvider().CommittedAPI().ProtocolParameters().Bech32HRP() != hrp {
+			return nil, ierrors.Wrapf(err, "invalid bech32 address prefix: %s", hrp)
+		}
+
+		accountAddr, ok := addr.(*iotago.AccountAddress)
+		if !ok {
+			return nil, ierrors.Errorf("invalid bech32 address, not an account: %s", ParamsValidator.AccountAddress)
+		}
+
+		return accountAddr, nil
+	}); err != nil {
+		return err
+	}
+
+	return c.Provide(func() (ed25519.PrivateKey, error) {
+		privateKeys, err := loadEd25519PrivateKeysFromEnvironment("VALIDATOR_PRV_KEY")
+		if err != nil {
+			return nil, ierrors.Errorf("loading validator private key failed, err: %w", err)
+		}
+
+		if len(privateKeys) == 0 {
+			return nil, ierrors.New("loading validator private key failed, err: no private keys given")
+		}
+
+		if len(privateKeys) > 1 {
+			return nil, ierrors.New("loading validator private key failed, err: too many private keys given")
+		}
+
+		privateKey := privateKeys[0]
+		if len(privateKey) != ed25519.PrivateKeySize {
+			return nil, ierrors.New("loading validator private key failed, err: wrong private key length")
+		}
+
+		return privateKey, nil
+	})
+}
+
+func run() error {
+	validatorAccount = NewEd25519Account(deps.AccountAddress.AccountID(), deps.PrivateKey)
+
+	executor = timed.NewTaskExecutor[TaskType](1)
+
+	return Component.Daemon().BackgroundWorker(Component.Name, func(ctx context.Context) {
+		Component.LogInfof("Starting Validator with IssuerID: %s", validatorAccount.ID())
+
+		checkValidatorStatus(ctx)
+
+		deps.NodeBridge.Events.LatestCommittedSlotChanged.Hook(func(details *nodebridge.Commitment) {
```

Comment:
Isn't that a bit of a race condition here?
Shouldn't you hook before checking the status? Could be that you miss an updated between those two lines.

Reply:
True, good catch!